engines/ic3base.cpp

```diff
@@ -346,32 +432,69 @@ bool IC3Base::rel_ind_check(size_t i, const IC3Formula & c, IC3Formula & out)
   solver_->assert_formula(solver_->make_term(Not, c.term));
   // Trans
   assert_trans_label();
-  // c'
-  solver_->assert_formula(ts_.next(c.term));
 
-  Result r = check_sat();
+  // use assumptions for c' so we can get cheap initial
+  // generalization if the check is unsat
+
+  // NOTE: relying on same order between assumps_ and c.children
+  assumps_.clear();
+  {
+    // TODO shuffle assumps and (a copy of) c.children
+    //      if random seed is set
+    Term lbl, ccnext;
+    for (const auto & cc : c.children) {
+      ccnext = ts_.next(cc);
+      lbl = label(ccnext);
+      solver_->assert_formula(solver_->make_term(Implies, lbl, ccnext));
+      assumps_.push_back(lbl);
+    }
+  }
+
+  Result r = check_sat_assuming(assumps_);
   if (r.is_sat()) {
-    if (options_.ic3_pregen_) {
-      out = generalize_predecessor(i, c);
-    } else {
-      out = get_model_ic3formula();
+    if (get_pred) {
+      if (options_.ic3_pregen_) {
+        out = generalize_predecessor(i, c);
+      } else {
+        out = get_model_ic3formula();
```

Comment:
`generalize_predecessor` is doing the same thing. No?

Reply:
Only by default, if it's overridden by the subclass then it might actually generalize.

Reply:
i see. then what about doing this `if (options_.ic3_pregen_)` check in generalize_predecessor when  we override in a derived class

Reply:
I don't like that as much because then every time you add a derived class you have to know to do that check. This way, it's only done once in the base class and you don't have to worry about it in the derived class.

Reply:
yeah. that is a good point.